sweet/benchmarks/cockroachdb/main.go

```diff
@@ -0,0 +1,611 @@
+}
+
+func reportFromBenchmarkOutput(b *driver.B, cfg *config, output string) (err error) {
+	defer func() {
+		if err != nil {
+			fmt.Fprintln(os.Stderr, "=== Benchmarking tool output ===")
+			fmt.Fprintln(os.Stderr, output)
+		}
+	}()
+
+	for _, metricType := range cfg.bench.metricTypes {
+		err = getAndReportMetrics(b, metricType, output)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+type benchmarkMetrics struct {
+	totalOps       uint64
+	opsPerSecond   uint64
+	averageLatency uint64
+	p50Latency     uint64
+	p95Latency     uint64
+	p99Latency     uint64
+	pMaxLatency    uint64
+}
+
+func getAndReportMetrics(b *driver.B, metricType string, output string) error {
+	metrics, err := getMetrics(metricType, output)
+	if err != nil {
+		return err
+	}
+	reportMetrics(b, metricType, metrics)
+	return nil
+}
+
+func getMetrics(metricType string, output string) (benchmarkMetrics, error) {
+	re := regexp.MustCompile(fmt.Sprintf(`.*(__total)\n.*%s`, metricType))
+	match := re.FindString(output)
+	if len(match) == 0 {
+		return benchmarkMetrics{}, fmt.Errorf("failed to find %s metrics in output", metricType)
+	}
+	match = strings.Split(match, "\n")[1]
+	fields := strings.Fields(match)
+
+	stringToUint64 := func(field string) (uint64, error) {
+		number, err := strconv.ParseFloat(field, 64)
+		if err != nil {
+			return 0, errors.Wrap(err, "Error parsing metrics to uint64")
+		}
+		return uint64(number), nil
+	}
+
+	uint64Fields := make([]uint64, len(fields[2:])-1)
+	for i := range uint64Fields {
+		var err error
+		uint64Fields[i], err = stringToUint64(fields[2+i])
+		if err != nil {
+			return benchmarkMetrics{}, err
+		}
+	}
+
+	metrics := benchmarkMetrics{
+		totalOps:       uint64Fields[0],
+		opsPerSecond:   uint64Fields[1],
+		averageLatency: uint64Fields[2],
+		p50Latency:     uint64Fields[3],
+		p95Latency:     uint64Fields[4],
+		p99Latency:     uint64Fields[5],
+		pMaxLatency:    uint64Fields[6],
+	}
+	return metrics, nil
+}
+
+func reportMetrics(b *driver.B, metricType string, metrics benchmarkMetrics) {
+	b.Report(fmt.Sprintf("%s-ops-sec", metricType), metrics.opsPerSecond)
+	b.Report(fmt.Sprintf("%s-ops-total", metricType), metrics.totalOps)
+	b.Report(fmt.Sprintf("%s-avg-latency-ms", metricType), metrics.averageLatency)
+	b.Report(fmt.Sprintf("%s-p50-latency-ms", metricType), metrics.p50Latency)
+	b.Report(fmt.Sprintf("%s-p95-latency-ms", metricType), metrics.p95Latency)
+	b.Report(fmt.Sprintf("%s-p99-latency-ms", metricType), metrics.p99Latency)
+	b.Report(fmt.Sprintf("%s-pMax-latency-ms", metricType), metrics.pMaxLatency)
+}
+
+func run(cfg *config) (err error) {
+	var instances []*cockroachdbInstance
+	// Launch the server.
+	instances, err = launchCockroachCluster(cfg)
+
+	if err != nil {
+		return fmt.Errorf("starting cluster: %v\n", err)
+	}
+
+	if err = waitForCluster(instances, cfg); err != nil {
+		return err
+	}
+
+	// Clean up the cluster after we're done.
+	defer func() {
```

Comment:
Methinks this `defer` should be moved up, otherwise return on L479 escapes; i.e., an error in `waitForCluster` would skip cleanup.

Reply:
Good catch, done.